src/System.Diagnostics.Process/src/System/Diagnostics/ProcessWaitState.Unix.cs

```diff
@@ -511,5 +489,78 @@ private Task WaitForExitAsync(CancellationToken cancellationToken = default(Canc
             });
         }
 
+        private bool TryReapChild()
+        {
+            lock (_gate)
+            {
+                if (_exited)
+                {
+                    return false;
+                }
+
+                // Try to get the state of the child process
+                int status;
+                int waitResult = Interop.Sys.WaitPid(_processId, out status, Interop.Sys.WaitPidOptions.WNOHANG);
+
+                if (waitResult == _processId)
+                {
+                    // Process has exited
+                    if (Interop.Sys.WIfExited(status))
+                    {
+                        _exitCode = Interop.Sys.WExitStatus(status);
+                    }
+                    else if (Interop.Sys.WIfSignaled(status))
+                    {
+                        const int ExitCodeSignalOffset = 128;
+                        _exitCode = ExitCodeSignalOffset + Interop.Sys.WTermSig(status);
+                    }
+
+                    SetExited();
+                    return true;
+                }
+                else if (waitResult == 0)
+                {
+                    // Process is still running
+                    return false;
+                }
+                else if (waitResult == -1)
+                {
+                    Debug.Fail("Unexpected errno value from waitpid");
+                }
+                else Debug.Fail("Unexpected process ID from waitpid.");
+
+                SetExited();
+                return true;
+            }
+        }
+
+        internal static void CheckChildren()
+        {
+            // This is called on SIGCHLD from a native thread.
+            // A lock in Process ensures no new processes are spawned while we are checking.
+            lock (s_childProcessWaitStates)
+            {
+                bool restart;
+                do
+                {
+                    restart = false;
+                    foreach (var kv in s_childProcessWaitStates)
+                    {
+                        ProcessWaitState pws = kv.Value;
+                        if (pws.TryReapChild())
+                        {
+                            bool removed = pws.ReleaseRef();
+                            // We have removed an item from s_childProcessWaitStates
+                            // so our iterator has become invalid.
```

Comment:
Maybe it's not much of a concern, but this could result in O(n^2) behavior.  An alternative is to instead do something like:
```C#
ProcessWaitState firstToRemove = null;
List<ProcessWaitState> additionalToRemove = null;
foreach (KeyValuePair<int, ProcessWaitState> kv in s_childProcessWaitStates)
{
    ProcessWaitState pws = kv.Value;
    if (pws.TryReapChild())
    {
        if (firstToRemove == null)
        {
            firstToRemove = pws;
        }
        else
        {
            if (additionalToRemove == null)
            {
                additionalToRemove = new List<ProcessWaitState>();
            }
            additionalToRemove.Add(pws);
        }
    }
}

if (firstToRemove != null)
{
    firstToRemove.ReleaseRef();
    if (additionalToRemove != null)
    {
        foreach (ProcessWaitState pws in additionalToRemove)
        {
            pws.ReleaseRef();
        }
    }
}
```

Reply:
Looks good!

One change I was going to look into is to call `waitpid(-1, &status, WNOWAIT|WNOHANG)`. This will give us a key we can use in the Dictionary. If we are unable to match that key, we should leave that child and fall back to looping.